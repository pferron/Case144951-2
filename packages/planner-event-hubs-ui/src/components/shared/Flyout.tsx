import React, { useEffect, useRef, useMemo, ReactNode } from 'react';
import { injectTestId } from '@cvent/nucleus-test-automation';
import useBreakpoints from '@hooks/useBreakpoints';
import Panel from '@cvent/carina/components/Panel';
import { CLICK, HOVER } from '@utils/constants';

const Keys = {
  ENTER: 'Enter',
  ESCAPE: 'Escape',
  TAB: 'Tab',
  ARROW_DOWN: 'ArrowDown',
  ARROW_UP: 'ArrowUp'
};

interface TriggerType {
  toggleOpen: () => void;
}

interface Props {
  trigger: ({ toggleOpen }: TriggerType) => ReactNode;
  children: ReactNode | ((...args) => ReactNode);
  isVisible: boolean;
  toggleVisible: (boolean) => void;
  triggerMethod: string;
  hasSpacerAfterTrigger: boolean;
}

/**
 * Component to display content -
 * 1. As a flyout below the trigger component for desktop view
 * 2. As a slide panel for mobile view
 */
function Flyout({
  trigger,
  children,
  isVisible,
  toggleVisible,
  triggerMethod = CLICK,
  hasSpacerAfterTrigger
}: Props): JSX.Element {
  const flyoutContainerRef = useRef(null);
  const flyoutRef = useRef();
  const toggleVisibleRef = useRef(toggleVisible);
  const { isMobile } = useBreakpoints();

  const handlers = useMemo(() => {
    // Memoize these handlers so the allowScroll/blockScroll effect can have one less dependency
    if (isMobile) {
      return {
        handleClickOutside: (event): void => {
          if (flyoutContainerRef?.current && !flyoutContainerRef?.current.contains(event.target)) {
            if (typeof toggleVisibleRef.current === 'function') toggleVisibleRef.current(false);
          }
        },
        onKeyDown: (event): void => {
          if (event.key === Keys.ESCAPE) {
            if (typeof toggleVisibleRef.current === 'function') toggleVisibleRef.current(false);
          }
        }
      };
    }
    return {
      handleClickOutside: (event): void => {
        if (flyoutContainerRef?.current && !flyoutContainerRef?.current.contains(event.target)) {
          toggleVisibleRef.current(false);
        }
      },
      onKeyDown: (event): void => {
        if (event.key === Keys.ESCAPE) {
          toggleVisibleRef.current(false);
        }
      }
    };
  }, [isMobile]);

  useEffect(() => {
    if (typeof window !== 'undefined') {
      if (isVisible) {
        document.addEventListener('click', handlers.handleClickOutside);
        document.addEventListener('keydown', handlers.onKeyDown);
      } else {
        document.removeEventListener('click', handlers.handleClickOutside);
        document.removeEventListener('keydown', handlers.onKeyDown);
      }
    }
  }, [isVisible, handlers.handleClickOutside, handlers.onKeyDown]);

  // Setting ref is technically a side-effect
  useEffect(() => {
    toggleVisibleRef.current = toggleVisible;
  }, [toggleVisible]);

  const component =
    typeof children === 'function' ? children({ toggleOpen: () => toggleVisible(!isVisible) }) : children;

  return (
    <div
      onMouseEnter={
        triggerMethod === HOVER
          ? (): void => {
              if (!isVisible) {
                toggleVisible(true);
              }
            }
          : null
      }
      onMouseLeave={triggerMethod === HOVER ? (): void => toggleVisible(false) : null}
      ref={flyoutContainerRef}
      {...injectTestId('flyout-container')}
    >
      <div>{trigger({ toggleOpen: () => toggleVisible(!isVisible) })}</div>
      {hasSpacerAfterTrigger ? <div /> : null}
      {isMobile ? (
        <div ref={flyoutRef}>
          <Panel elevated isOpen={isVisible} testID="mobile-navigation-panel">
            {component}
          </Panel>
        </div>
      ) : (
        <div {...injectTestId('navigation-panel')}>
          <div ref={flyoutRef}>{component}</div>
        </div>
      )}
    </div>
  );
}

export default Flyout;
