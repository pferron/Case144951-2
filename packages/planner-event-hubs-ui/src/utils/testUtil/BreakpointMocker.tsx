import React from 'react';
import { BreakpointsContext } from '@cvent/carina/components/Breakpoints/BreakpointsProvider';

const defaultBreakpoints = { s: 336, m: 560, l: 1080, xl: 1464 };

export function BreakpointMocker({
  children,
  isS = false,
  isDefaultSize = false,
  isM = false,
  isL = false,
  isXL = false
}: Props): JSX.Element {
  // FIREBALL
  // eslint-disable-next-line react/jsx-no-constructed-context-values
  const req = {
    isS,
    isDefaultSize,
    isM,
    isL,
    isXL,
    ...defaultBreakpoints
  };
  return <BreakpointsContext.Provider value={req}>{children}</BreakpointsContext.Provider>;
}

interface Props {
  children: JSX.Element;
  isDefaultSize?: boolean;
  isS?: boolean;
  isM?: boolean;
  isL?: boolean;
  isXL?: boolean;
}
