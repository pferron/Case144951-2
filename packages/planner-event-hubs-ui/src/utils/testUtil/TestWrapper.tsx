import React, { useMemo } from 'react';
import { LocalizationProvider } from 'nucleus-text';
import { Breakpoints } from '@cvent/carina/components/Breakpoints';
import { AppFeature } from '@cvent/planner-event-hubs-model/types';
import { ThemeProvider } from '@cvent/carina/components/ThemeProvider/ThemeProvider';
import { getDefaultTheme } from '@cvent/carina/utils/tokens';
import { AppFeaturesProvider } from '@components/AppFeaturesProvider';
import { UserPermissionsContext } from '@hooks/UserPermissionsProvider';
import locales, { defaultLocale } from '../../../locales';

export function TestWrapper({ children, appFeatures }: Props): JSX.Element {
  const userPermissionsValue = useMemo(() => ({ VideoCenter: true }), []);
  return (
    <ThemeProvider theme={getDefaultTheme()}>
      <LocalizationProvider locale={defaultLocale} locales={locales}>
        <Breakpoints ssrSize="m" key={typeof window === 'undefined' ? 'ssr' : 'csr'}>
          <AppFeaturesProvider value={appFeatures}>
            <UserPermissionsContext.Provider value={userPermissionsValue}>{children}</UserPermissionsContext.Provider>
          </AppFeaturesProvider>
        </Breakpoints>
      </LocalizationProvider>
    </ThemeProvider>
  );
}

interface Props {
  children: JSX.Element;
  appFeatures?: AppFeature[];
}
