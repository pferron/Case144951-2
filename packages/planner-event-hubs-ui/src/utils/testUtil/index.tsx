import { render, configure } from '@testing-library/react';
import { TestWrapper } from './TestWrapper';

configure({ testIdAttribute: 'data-cvent-id' });

type RenderParams = Parameters<typeof render>;
type RenderReturn = ReturnType<typeof render>;
const customRender = (ui: RenderParams[0], options?: RenderParams[1]): RenderReturn =>
  render(ui, { wrapper: TestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
