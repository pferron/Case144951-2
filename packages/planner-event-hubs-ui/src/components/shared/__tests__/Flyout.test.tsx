import React, { ReactNode, useState } from 'react';
import { fireEvent, render, screen } from '@utils/testUtil';
import { injectTestId } from '@cvent/nucleus-test-automation';
import { BreakpointsContext } from '@cvent/carina/components/Breakpoints/BreakpointsProvider';
import PropTypes from 'prop-types';
import Flyout from '../Flyout';

const size = { s: 0, m: 0, l: 0, xl: 0, isS: true };

function Wrapper({ hoverTrigger = false }): JSX.Element {
  const [isVisible, setVisibility] = useState(false);
  const trigger = ({ toggleOpen }): ReactNode => (
    <div {...injectTestId('flyout-trigger')} onClick={toggleOpen} onKeyDown={() => null} role="button" tabIndex={0}>
      Open flyout
    </div>
  );

  return (
    <Flyout
      isVisible={isVisible}
      toggleVisible={setVisibility}
      trigger={trigger}
      triggerMethod={hoverTrigger ? 'HOVER' : undefined}
      hasSpacerAfterTrigger
    >
      {(): ReactNode => <div>Flyout content</div>}
    </Flyout>
  );
}
Wrapper.propTypes = {
  hoverTrigger: PropTypes.bool
};

describe('Flyout', () => {
  describe('for desktop view', () => {
    it('should open flyout initially', async () => {
      const { container } = render(<Wrapper />);
      expect(container).toMatchSnapshot();
      expect(await screen.findByTestId('flyout-container')).toBeInTheDocument();
      expect(screen.queryByText('Flyout content')).toBeVisible();
    });

    it('should open flyout after click', async () => {
      render(<Wrapper />);
      expect(screen).toMatchSnapshot();
      expect(await screen.findByTestId('flyout-trigger')).toBeInTheDocument();
      fireEvent.click(screen.getByTestId('flyout-trigger'));
      expect(screen.getByText('Flyout content')).toBeVisible();
    });
    it('should open flyout after mouseEnter', async () => {
      render(<Wrapper hoverTrigger />);
      expect(screen).toMatchSnapshot();
      expect(await screen.findByTestId('flyout-container')).toBeInTheDocument();
      const element = await screen.findByTestId('flyout-container');
      fireEvent.mouseEnter(element);
      expect(await screen.findByText('Flyout content')).toBeVisible();
    });
  });

  describe('for mobile view', () => {
    it('should not open flyout initially', async () => {
      render(
        <BreakpointsContext.Provider value={size}>
          <Wrapper />
        </BreakpointsContext.Provider>
      );

      expect(await screen.findByTestId('flyout-container')).toBeInTheDocument();
      expect(screen.queryByTestId('mobile-navigation-panel')).not.toBeInTheDocument();
    });

    it('should open flyout after click', async () => {
      render(
        <BreakpointsContext.Provider value={size}>
          <Wrapper />
        </BreakpointsContext.Provider>
      );

      expect(await screen.findByTestId('flyout-trigger')).toBeInTheDocument();
      fireEvent.click(screen.getByTestId('flyout-trigger'));
      expect(await screen.findByTestId('mobile-navigation-panel')).toBeInTheDocument();
      expect(screen.getByText('Flyout content')).toBeInTheDocument();
    });
  });
});
