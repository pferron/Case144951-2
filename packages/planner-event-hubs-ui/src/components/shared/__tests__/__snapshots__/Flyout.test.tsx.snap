// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flyout for desktop view should open flyout after click 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Flyout for desktop view should open flyout after mouseEnter 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Flyout for desktop view should open flyout initially 1`] = `
<div>
  <div
    data-cvent-id="flyout-container"
  >
    <div>
      <div
        data-cvent-id="flyout-trigger"
        role="button"
        tabindex="0"
      >
        Open flyout
      </div>
    </div>
    <div />
    <div
      data-cvent-id="navigation-panel"
    >
      <div>
        <div>
          Flyout content
        </div>
      </div>
    </div>
  </div>
</div>
`;
